src/sources/util/tcp/request_limiter.rs

```diff
@@ -0,0 +1,158 @@
+use std::cmp::Ordering;
+use std::sync::{Arc, Mutex};
+use tokio::sync::{OwnedSemaphorePermit, Semaphore};
+
+const EWMA_WEIGHT: f32 = 0.1;
+
+const MINIMUM_PERMITS: usize = 2;
+
+pub struct RequestLimiterPermit {
+    semaphore_permit: Option<OwnedSemaphorePermit>,
+    request_limiter_data: Arc<Mutex<RequestLimiterData>>,
+}
+
+impl RequestLimiterPermit {
+    pub fn decoding_finished(&self, num_events: usize) {
+        let mut request_limiter_data = self.request_limiter_data.lock().unwrap();
+        request_limiter_data.update_average(num_events);
+    }
+}
+
+impl Drop for RequestLimiterPermit {
+    fn drop(&mut self) {
+        if let Ok(mut request_limiter_data) = self.request_limiter_data.lock() {
+            let target = request_limiter_data.target_requests_in_flight();
+            let current = request_limiter_data.total_permits;
+
+            match target.cmp(&current) {
+                Ordering::Greater => request_limiter_data.increase_permits(),
+                Ordering::Equal => {
+                    // only release the current permit (when the inner permit is dropped automatically)
+                }
+                Ordering::Less => {
+                    let permit = self.semaphore_permit.take().unwrap();
+                    request_limiter_data.decrease_permits(permit);
+                }
+            }
+        }
+    }
+}
+
+struct RequestLimiterData {
+    event_limit_target: usize,
+    total_permits: usize,
+    average_request_size: f32,
+    semaphore: Arc<Semaphore>,
+    max_requests: usize,
+}
+
+impl RequestLimiterData {
+    pub fn update_average(&mut self, num_events: usize) {
+        let num_events = num_events as f32;
+        self.average_request_size =
+            (EWMA_WEIGHT * num_events) + ((1.0 - EWMA_WEIGHT) * self.average_request_size);
```

Comment:
Note that we already have an EWMA abstraction in `src/stats.rs`.

Reply:
Good to know, I'll use that one.

Reply:
I actually ended up adding yet another EWMA abstraction in `src/stats.rs`. The existing one had an optional initial value, but I wanted to initialize it so that the number of requests allowed was artificially low until a sufficient number of samples was obtained.